### DISCUSSION TITLE
A warning about freezegun on full unit test classes

### DISCUSSION THREAD

**Discussion by yajo** [OCA/odoo-community.org#122] [General]
I just wanted to warn the wider community about a little corner problem I found recently with freezegun.

There's definitely something weird in [the way it patches `setUpClass` or other classmethods](https://github.com/spulec/freezegun/blob/4f4496380deefceead7bef23bccaca17c2bdecfa/freezegun/api.py#L572-L608).

In short, the class-level cleanups are bound to a different in-memory object. This produces that a class decorated with `@freezegun.freeze_time(some_date)` losses class cleanups. A particularly important one is [the one that closes the cursor](https://github.com/odoo/odoo/blob/f025224ccb9bef690f6a171fafcde039ab2be02b/odoo/tests/common.py#L708). Thus, these classes leave a dangling cursor after exercising all tests within.

This can be or not a problem, depending on what happens after the tests are run. If Odoo just finishes, no problem; the cursor will get reaped by other means. But if you keep testing other modules whose installation means a DB lock in some way, you'll get to clunky, hard-to-diagnose Postgres locks.

This was notified upstream in https://github.com/spulec/freezegun/issues/485 but:
- Don't be fooled by the title. It does affect other python versions too.
- There's zero activity to fix that.
- Odoo still requires very old freezegun versions in their `requirements.txt` files.

So, the most sensible approach for us here downstream at OCA, for now, is to replace such decorators like I did on https://github.com/OCA/sale-workflow/pull/2747. 

Take this example as a rule of thumb:

```diff
-@freeze_time("2021-10-02 15:30:00")
 class ExampleCase(TransactionCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        # HACK https://github.com/spulec/freezegun/issues/485
+        freezer = freeze_time("2021-10-02 15:30:00")
+        freezer.__enter__()
+        cls.addClassCleanup(freezer.__exit__)
```

Thanks!

**Comment by simahawk** [comment]
Interesting! What about decorating test methods? (Which btw is better from my POV in any case :wink: )

**Reply by yajo** [reply]
That works fine, but if you're creating frozen data in setUpClass, it won't work (unless you decorate that method too).
